### ISSUE TITLE
Unicode font do not display properly in map.

### ISSUE THREAD

**Issue by aerotypebharat** [maplibre/maplibre-gl-js#50]
This is the name in openstreetmap.org 
![image](https://user-images.githubusercontent.com/10284102/103306225-b004da80-49d2-11eb-9dfc-cf48f34d2de8.png)
This is what displayed in map. 
![image](https://user-images.githubusercontent.com/10284102/103306313-e3476980-49d2-11eb-8c42-5865024b0137.png)
Test code available in code pen here
https://codepen.io/snickell/pen/dypOWzj

**Comment by aerotypebharat** [comment]
I opened this issue in Mapbox gl js. Can anyone give me direction how this can be solved to support Devanagari letters

**Comment by HarelM** [comment]
I'm not sure I fully understand the problem. It would be best if you elaborate.
Regardless, you'll probably need to download a font that support these letters and use it instead of the current font.
I had the same problem with hebrew letters I think...

**Comment by aerotypebharat** [comment]
![image](https://user-images.githubusercontent.com/10284102/126075804-f0d7758d-a92c-4fb3-864b-57c610179a96.png)
This vowel character is supposed to go in front of the consonant. 
![image](https://user-images.githubusercontent.com/10284102/126075868-fbb44a12-7b7c-43aa-a985-e356d6caa05f.png)

**Comment by HarelM** [comment]
Ahh, I have the same problem with "nikud" - the hebrew extra characters to indicate how to read, they appear after the letter instead of within the letter.
I have no clue how to solve this, you'll need to research how text in webgl is created and how to add these letters in the right place.
Sorry I don't have better knowledge on the subject. But if you find out please serve a PR :-)

**Comment by ambientlight** [comment]
@aerotypebharat: What fontface is used in your screenshot? Do you have a public fontface that renders Devanagari correctly at https://fonts.google.com/ ? You can then use https://github.com/orangemug/font-glyphs to generate prerendered fonts (Signed Distance Field glyphs) to consume from your style

**Comment by aerotypebharat** [comment]
I am using Khula font. It is public in https://fonts.google.com/. It renders properly in text. I have tried other fonts that support Devanagari but maplibre-gl.js does not have code to handle vowels and special characters of Devanagari.

**Comment by anarchodin** [comment]
The approach chosen by Mapbox to render labels is unfortunately fundamentally incompatible with scripts that require more sophisticated handling than a mechanical typewriter could provide. This is not a problem with Unicode support as such; there is no problem with the encoding.

**Comment by nyurik** [comment]
There was some attempts to solve this, but they didn't go far. OpenLayers tried to do it using svg on canvas, but iirc it was not solving it properly either. 

Basically there is a c++ lib that knows how to do all the proper text rendering.  Browsers use that for the sites. Webgl has limited support of that lib (?), eg cannot use it along a curve to draw street names. So mb had to create a separate system for fonts that can draw labels in webgl along a curve line, but only if each letter is by itself. It does not support ligatures and other complexities.  They did a hack for Arabic by creating 4 variants for each letter:  without ligatures, with one on left, one in right, and both, plus some code that decides which variant to use.  Languages such as used in Maianmar (Birma) are far more complex, and cannot be hacked this way. 

Hopefully some day browsers will have a good font support in webgl so all this hackary can just be dropped

**Comment by HarelM** [comment]
Assigned XXL bounty.
https://github.com/maplibre/maplibre/issues/193
Acceptance criteria: brahmic fonts are rendered correctly.

**Comment by hami9x** [comment]
@HarelM I'd like to work on this issue, can I claim this?

**Comment by hami9x** [comment]
@wipfli Thank you for the heads up, I have some background in C++ and enjoy hardcore stuff like this. This will be really challenging though.

**Comment by HarelM** [comment]
It's worth considering writing it in Rust maybe, so that it will be available in both native and here (as wasm).

**Comment by bigBxr** [comment]
The arabic is displaying backwards and the letters are disconnected.

**Comment by wipfli** [comment]
@bigBxr to get correct arabic text you need to load the RTL plugin

**Comment by swastikCommits** [comment]
Is this still open to work with bounty?

**Comment by HarelM** [comment]
Definitely, but please make sure to coordinate with @wipfli and @1ec5 as I belive they are working on it using some grant.

**Comment by 1ec5** [comment]
@swastikCommits do you have an idea of how you’d solve this issue? What I’m working on is outlined in maplibre/maplibre-style-spec#1045. The approach I’m taking won’t address the issue 100%, but it’ll come close, addressing most of the cases that have been reported so far. If you have an alternative or complementary approach that you can make a compelling case for, maybe there’s room for it too.

**Comment by wipfli** [comment]
I am not working on correct text rendering at the moment. But you can use the maplibre-gl-complex-text plugin to render most Indic scripts.

**Comment by swastikCommits** [comment]
@1ec5 @HarelM currently the Devanagari font is taken care by the server side glyphs instead of rendering it locally. So we don't want to render it using the local glyph rendering using TinySDF? Am I right?

As I see Devanagari being commented out here:

![Image](https://github.com/user-attachments/assets/0085fa63-e7be-459e-a9a5-ba28de2ba69e)


TinySDF doesn't handle complex script shaping (combining characters, ligatures). Proper Devanagari rendering requires special handling of conjuncts and half-forms. So, for proper Devanagari support, MapLibre would need a more sophisticated text shaping engine like HarfBuzz.

An alterante approach would be:
- Create an HTML overlay with properly rendered Devanagari text
- Position it over the map at the correct coordinates
- Update its position during map movements
(as I believe the symbol is only breaking for a single character, rest other Devanagari strings are working fine)

**Comment by HarelM** [comment]
HTML can already be done now using a `Marker`, but when drawing enough HTML elements the browser chokes, so that isn't a good solution I believe...

**Comment by swastikCommits** [comment]
Yeah you're right about that. It would the worst case to handle the rendering.
Hence, I have to deep dive upon how the MapLibre is handling the glpyh responses from the server

**Comment by HarelM** [comment]
It's worth mentioning that this is also available:
https://www.npmjs.com/package/maplibre-local-glyphs
Which I'm not sure how relevant it is to this issue, but this is discussed hereto some extent:
https://github.com/maplibre/maplibre-gl-js/discussions/2555#discussioncomment-12340407

**Comment by swastikCommits** [comment]
Thanks for mentioning the glyph package. I'm currently working on it and will soon raise a PR

**Comment by louwers** [comment]
There is an open [style spec proposal](https://github.com/maplibre/maplibre-style-spec/pull/1184) to specify font faces in a style. There is a corresponding [PR](https://github.com/maplibre/maplibre-native/pull/1439) that integrates HarfBuzz into Native for text shaping.

> It's worth considering writing it in Rust maybe, so that it will be available in both native and here (as wasm).

Note that [HarfRust](https://github.com/harfbuzz/harfrust) is a thing now, but since it is 3x slower than the C++ version right now, Native will not use it until it gets more performant.

**Comment by 1ec5** [comment]
> TinySDF doesn't handle complex script shaping (combining characters, ligatures). Proper Devanagari rendering requires special handling of conjuncts and half-forms. So, for proper Devanagari support, MapLibre would need a more sophisticated text shaping engine like HarfBuzz.

The approach I’m taking is a stopgap until we can depend on HarfBuzz everywhere. It relies on TinySDF, which _can_ render Devanagari correctly via the platform’s text layout engine – as long as we avoid segmenting the text in the wrong places. So the main thing I’m doing is changing the segmentation logic, and removing the requirement that a glyph represent a single Unicode codepoint, something any HarfBuzz-based approach would need anyways.

**Comment by 1ec5** [comment]
I see that the style specification compatibility table for [`font-faces`](https://maplibre.org/maplibre-style-spec/font-faces/) links to this issue to track adding support to GL JS. But this issue is about the broader goal of supporting complex writing systems. Realistically this will require a combination of multiple approaches. All of these approaches have advantages and disadvantages depending on the developer’s needs. Should we break out a separate issue to track a `font-faces` implementation specifically?

**Comment by HarelM** [comment]
Isn't font-faces the last thing needed to solve this issue? 
If not, it might be beneficial to have this issue as a collection of issues that needs to be solved in order to address this requirement. 
I.e. make this issue a "super issue".

**Comment by HarelM** [comment]
Oh, yeah, complex shaping is still needed. 
Feel free to open a different issue to track font-faces.

**Comment by 1ec5** [comment]
#6637 now tracks `font-faces` support specifically.

**Comment by 1ec5** [comment]
This is blocked by node-unicode/node-unicode-data#57 unless we’re OK with departing from the approach in #6639 and rolling our own script for fetching Indic character property data directly from the Unicode Character Database.

**Comment by 1ec5** [comment]
There are two remaining aspects of complex text layout that we need to implement:

* [x] Font substitution: #4564
* [x] Astral planes: #6640
* [ ] Ligatures and combining characters: #2384 #6232
* [ ] Contextual forms: mapbox/mapbox-gl-js#4007
